app/domain/authentication/authn_k8s/validate_pod_request.rb

```diff
@@ -13,16 +13,16 @@ module AuthnK8s
         validate_webservice_is_whitelisted:  ::Authentication::Security::ValidateWebserviceIsWhitelisted.new,
         validate_role_can_access_webservice: ::Authentication::Security::ValidateRoleCanAccessWebservice.new,
         enabled_authenticators:              Authentication::InstalledAuthenticators.enabled_authenticators_str(ENV),
-        validate_resource_restrictions:      ValidateResourceRestrictions.new,
-        extract_container_name: ExtractContainerName.new,
-        logger: Rails.logger
+        validate_resource_restrictions:      Authentication::ResourceRestrictions::ValidateResourceRestrictions.new(
+          extract_resource_restrictions:      ExtractK8sResourceRestrictions.new
+        ),
```

Comment:
was this a thing in other authenticators? a dependency that couples another one?

Reply:
The command class dependencies meant to be able for replacement.
This is how you override its internal behavior, by replacing its dependency with another one that uses the same inputs.
This is how you implement hierarchy in command-class.

Reply:
How would you replace a dependency in a command class otherwise?

Reply:
@abrahamko Why `ValidateResourceRestrictions` command class contains 1 dependency instead of 3?
You dont think in the future we will need this capability of replacing the logic of `validate_resource_restrictions_configuration , validate_request_matches_resource_restrictions` like you did here for the `extract_resource_restrictions` logic ?

Reply:
The other 2 already differ for each authenticator.
1. Validate restrictions is done using the `@constraints` input
2. Validate request is done using `AuthenticationRequest` by calling `valid_restriction?` on each restriction.
The code around them is generic, so I don't expect it to change.

Reply:
I feel comfortable the way that is, as this is the way to replace the default dependency.
I think this is a correct way to do it.

Reply:
@jonahx can you take a look on this? it seems odd to me that we have `app` code using dependencies. I understand the need here but want to have another look on it.

Reply:
> it seems odd to me that we have app code using dependencies. 

@orenbm Can you clarify your concern here?  

To me it looks like a nice example of re-use by composition, but I'm curious what in particular you're worried about.

Admittedly the total _number_ of dependencies to this class is a little concerning, but it seems to be the "top level" validation class, just delegating out to all the individual validations, and since we have a lot of validations to do (just because of our business logic), I think it's ok.  The CommandClass is doing what it's supposed to do by making all of our dependencies stare us in the face.

Reply:
no real concern. I just remember you saying that inputs will be given at run time in the app code and dependencies will be given only in UTs. if this is not the case then i'm good.

Just to verify we're looking at the same thing - i am talking about the fact that `validate_resource_restrictions` is initialized as a dependency and replaces a dependency (`extract_resource_restrictions`) of `Authentication::ResourceRestrictions::ValidateResourceRestrictions` in its initialization.

I'm not against it but it is the first time that this happens in our code (as far as i know)

Reply:
>  I just remember you saying that inputs will be given at run time in the app code and dependencies will be given only in UTs. if this is not the case then i'm good.

Ah, so just a slight miscommunication there. Yes: inputs are only given at runtime.  Yes: dependencies are given at "compile time" -- I think of it like that, even though ruby has no compilation.  It just means: dependencies are known before any code runs.  

> I'm not against it but it is the first time that this happens in our code (as far as i know)

Often, the default dependency is the only one we need, other than in our UTs.  So UTs are a common place to see the default dep changed out.  However, they need not be the only place.  And in fact, the ability to swap them out is one of the intended benefits of this style.

As I understand the code, `validate_resource_restrictions` defines some high-level steps:

```
        extract_resource_restrictions
        validate_resource_restrictions_configuration
        validate_request_matches_resource_restrictions
```

What each _specific_ step means might vary from one authenticator to another, and because the steps are just composed of other objects, we can change them out to suit individual authenticators, and keep the same high-level logic, without being forced to repeat it.  This is textbook "composition over inheritance" and I give it a 👍.

Reply:
great, thanks for clarifying!